### PR TITLE
[Bridge] Checkpoint

### PR DESCRIPTION
**Pull request by Kourin1996**
Fix Edge-459

# Description

This PR is a base PR for Bridge Checkpoint, checked out from `bridge` branch.

TODO for Phase 1
- [x] Create bridge/checkpoint package bridge
- [x] Generate new Checkpoint at the beginning of every epoch
- [x] Create Transport for Checkpoint Gossip
- [x] Sign generated Checkpoint and gossip it
- [x] Show logs once collected 2/3+ signatures

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [ ] I have tested this code manually

### Manual tests

Please complete this section if you ran manual tests for this functionality, otherwise delete it

# Documentation update

Please link the documentation update PR in this section if it's present, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: bridge/checkpoint/checkpoint.go
Comment (0xAleksaOpacic) [inline bridge/checkpoint/checkpoint.go]: What value we want to have for checkpoint EpochSize?
1. Some default value
2. Same as epoch size for consensus
3. Cli config to setup + default value

File: bridge/checkpoint/checkpoint.go
Comment (0xAleksaOpacic) [inline bridge/checkpoint/checkpoint.go]: When are we calling this method? Is it from ibft insertBlock, on every **checkpoint new epoch**?

File: bridge/checkpoint/checkpoint.go
Comment (0xAleksaOpacic) [inline bridge/checkpoint/checkpoint.go:79]: If we know what is currentEpoch in the root chain, do we need the last-child block? 
We can use currentEpoch*epochSize + 1 for lower bound

File: bridge/checkpoint/checkpoint.go
Comment (0xAleksaOpacic) [inline bridge/checkpoint/checkpoint.go]: Is there a chance that some epoch is not going to get inserted in the rootchain contract?
If yes, and let's say our chain is on epoch 7, and the rootchain has been last updated on epoch 3. 
Will our range be from epoch 3 to latest block, or only one epoch (3-4)?

**Review by 0xAleksaOpacic — APPROVED** [review]
Great job, I have left you a few questions so I can understand structure better, overall looks good

File: bridge/checkpoint/checkpoint.go
Comment (0xAleksaOpacic) [reply bridge/checkpoint/checkpoint.go]: @Kourin1996 Ping

File: bridge/checkpoint/checkpoint.go
Comment (0xAleksaOpacic) [reply bridge/checkpoint/checkpoint.go]: @Kourin1996 Ping

File: bridge/checkpoint/checkpoint.go
Comment (0xAleksaOpacic) [reply bridge/checkpoint/checkpoint.go:79]: @Kourin1996 Ping

File: bridge/checkpoint/checkpoint.go
Comment (Kourin1996) [reply bridge/checkpoint/checkpoint.go]: @Aleksao998 As discussed yesterday, must be No.3 and Epoch size can be given from CLI config or genesis.json

File: bridge/checkpoint/checkpoint.go
Comment (Kourin1996) [reply bridge/checkpoint/checkpoint.go]: @Aleksao998 on every checkpoint new epoch, which means should call when IBFT consensus begins for block #201 if next checkpoint is for blocks from 100 and 200

File: bridge/checkpoint/checkpoint.go
Comment (Kourin1996) [reply bridge/checkpoint/checkpoint.go:79]: @Aleksao998 I created mock for phase 1. Actually we need to know last child block because no guarantee that every checkpoint is created from fixed number of blocks

Comment (0xAleksaOpacic) [reply bridge/checkpoint/checkpoint.go:79]: Thank you for an explanation!

File: bridge/checkpoint/checkpoint.go
Comment (Kourin1996) [reply bridge/checkpoint/checkpoint.go]: Yes that's why we'll have Ack and NoAck mechanism

File: bridge/checkpoint/checkpoint.go
Comment (0xAleksaOpacic) [reply bridge/checkpoint/checkpoint.go]: Thank you!

File: bridge/checkpoint/checkpoint.go
Comment (0xAleksaOpacic) [reply bridge/checkpoint/checkpoint.go]: Agree, @lazartravica you said we should go for the same value (consensus epochSize = checkpoint epochSize), but I agree on this with Yoshiki. What do you think?

File: bridge/checkpoint/checkpoint.go
Comment (0xAleksaOpacic) [reply bridge/checkpoint/checkpoint.go]: So checkpoint range should be [fromLastChildBlock, toLastBlockInPolygonEdge].

Comment (Kourin1996) [reply bridge/checkpoint/checkpoint.go]: Ideally

File: bridge/checkpoint/checkpoint.go
Comment (Kourin1996) [reply bridge/checkpoint/checkpoint.go]: @Aleksao998 We can use same value but default IBFT epoch size is too big

Comment (0xAleksaOpacic) [reply bridge/checkpoint/checkpoint.go]: That's why I agree with you that we should have separate flag.

**Review by 0xAleksaOpacic — APPROVED** [review]
LGTM! Once we merge all 3 PR into this branch we would need to review it again.

**Comment by zivkovicmilos** [comment]
Closing this PR as it's been archived.

**Pull request closed without merging**
